### PR TITLE
route review cursor and system caret

### PR DESCRIPTION
**Pull request by burmancomp**
### Link to issue number:
14885
### Summary of the issue:
It has not been possible to move system focus/caret with single routing button press when braille is tethered to review, or when braille is tethered to automatically and follows temporarily review cursor. When user wants to review editable text without affecting system caret position (braille tethered to review continuously or temporarily), it maybe useful if he/she can route caret when needed for example to correct typos. Although thiskinds of routing has been possible, it has required at least one extra routing button press.
### Description of user facing changes
There is new setting "Route review cursor and system caret" in braille category. Setting is disabled by default. User can also add input gesture. If braille is tethered to focus, setting is hidden and gesture reports itself as unavailable.

If setting is enabled, and braille is tethered to review, or to automatically and braille follows temporarily review, routing buttons try to move system focus/caret.
### Description of development approach
This is modification from navigatorObject_moveFocus script so results should be similar.
### Testing strategy:
Tested using Albatross 80 model.
### Known issues with pull request:
none
### Change log entries:
New features
It is possible to move system caret with routing buttons especially in edit areas when braille is tethered to review.
Changes
Bug fixes
For Developers

### Code Review Checklist:

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] Security precautions taken.

### CODE REVIEW

File: source/globalCommands.py [hunk 1]
```diff
@@ -3315,6 +3316,45 @@ def script_braille_cycleShowSelection(self, gesture: inputCore.InputGesture) ->
 			msg = _("Braille show selection %s") % BoolFlag[nextName].displayString
 		ui.message(msg)
 
+	@script(
+		# Translators: Input help mode message for cycle through
+		# braille route review cursor and system caret command.
+		description=_("Cycle through the braille route review cursor and system caret states"),
```
Comment (leonardder) [inline source/globalCommands.py:3322]: I think this needs rewording, though I'm not yet sure how.

File: source/globalCommands.py [hunk 2]
```diff
@@ -3315,6 +3316,45 @@ def script_braille_cycleShowSelection(self, gesture: inputCore.InputGesture) ->
...
 		ui.message(msg)
 
+	@script(
+		# Translators: Input help mode message for cycle through
+		# braille route review cursor and system caret command.
+		description=_("Cycle through the braille route review cursor and system caret states"),
+		category=SCRCAT_BRAILLE
+	)
+	def script_braille_cycleRouteReviewCursorAndSystemCaret(self, gesture: inputCore.InputGesture) -> None:
+		"""If braille is not tethered to focus, set next state of braille route review cursor and system caret.
```
Comment (leonardder) [inline source/globalCommands.py:3326]: Note that this doc string gets lost after decoration. Better make this generic comment style.

File: source/globalCommands.py [hunk 3]
```diff
@@ -3315,6 +3316,45 @@ def script_braille_cycleShowSelection(self, gesture: inputCore.InputGesture) ->
...
+		State is reported using ui.message.
+		"""
+		# Unavailable if tether is to focus.
+		if TetherTo.FOCUS.value == config.conf["braille"]["tetherTo"]:
+			ui.message(
+				_(
+					# Translators: Gesture is unavailable because braille tether is to focus.
+					"Gesture is unavailable because braille tether is to focus"
```
Comment (leonardder) [inline source/globalCommands.py:3334]:
```suggestion
					"Action unavailable. Braille is tethered to focus"
```

File: source/globalCommands.py [hunk 4]
```diff
@@ -3483,8 +3523,74 @@ def script_braille_scrollForward(self, gesture):
 		description=_("Routes the cursor to or activates the object under this braille cell"),
 		category=SCRCAT_BRAILLE
 	)
-	def script_braille_routeTo(self, gesture):
+	def script_braille_routeTo(self, gesture: inputCore.InputGesture) -> None:
+		"""Routes cursor to or activates the object under this braille cell."""
```
Comment (leonardder) [inline source/globalCommands.py:3527]: Doc string gets lost when decorating.

```suggestion
		# Routes cursor to or activates the object under this braille cell.
```

File: source/globalCommands.py [hunk 5]
```diff
@@ -3483,8 +3523,74 @@ def script_braille_scrollForward(self, gesture):
 		description=_("Routes the cursor to or activates the object under this braille cell"),
 		category=SCRCAT_BRAILLE
 	)
-	def script_braille_routeTo(self, gesture):
+	def script_braille_routeTo(self, gesture: inputCore.InputGesture) -> None:
+		"""Routes cursor to or activates the object under this braille cell."""
+		# Try to ensure that braille message can be dismissed.
+		if braille.handler.buffer is braille.handler.messageBuffer:
```
Comment (leonardder) [inline source/globalCommands.py:3529]: Please move this below the call of `braille.handler.routeTo(gesture.routingIndex)` so you no longer have to duplicate that line.

File: source/globalCommands.py [hunk 6]
```diff
@@ -3483,8 +3523,74 @@ def script_braille_scrollForward(self, gesture):
...
+			braille.handler.routeTo(gesture.routingIndex)
+			return
 		braille.handler.routeTo(gesture.routingIndex)
+		# To proceed braille should be tethered to review cursor which means
+		# that tether is automatic (and review cursor is shown in braille)
+		# or tether is to review cursor.
+		# "Route review cursor and system caret" should be also enabled
+		# in braille settings.
+		if (
+			TetherTo.REVIEW.value != config.conf["braille"]["tetherTo"]
```
Comment (leonardder) [inline source/globalCommands.py:3539]: This puzzles me. This is not going to work when tethering automatically, right?

```suggestion
			self._tether != TetherTo.REVIEW.value
```

File: source/globalCommands.py [hunk 7]
```diff
@@ -3483,8 +3523,74 @@ def script_braille_scrollForward(self, gesture):
...
+			or not config.conf["braille"]["routeReviewCursorAndSystemCaret"]
+		):
+			return
+		navigatorObject: NVDAObject = api.getNavigatorObject()
+		if not isinstance(navigatorObject, NVDAObject):
+			ui.message(
+				_(
+					# Translators: There is no focusable object e.g. cannot use
+					# tab and shift tab to move to controls.
+					"No focus"
```
Comment (leonardder) [inline source/globalCommands.py:3549]: I don't think cursor routing should have spoken feedback like this.

File: source/globalCommands.py [hunk 8]
```diff
@@ -3483,8 +3523,74 @@ def script_braille_scrollForward(self, gesture):
...
+				)
+			)
+			return
+		# Try to set focus to current navigator object
+		if navigatorObject != api.getFocusObject():
+			# This script is available on the lock screen via getSafeScripts, as such
+			# ensure the navigatorObject does not contain secure information
+			# before setting focus to this object
+			if objectBelowLockScreenAndWindowsIsLocked(navigatorObject):
+				ui.message(gui.blockAction.Context.WINDOWS_LOCKED.translatedMessage)
```
Comment (leonardder) [inline source/globalCommands.py:3559]: Avoid messages in this script.

File: source/globalCommands.py [hunk 9]
```diff
@@ -3483,8 +3523,74 @@ def script_braille_scrollForward(self, gesture):
...
+				navigatorObject.setFocus()
+			except COMError:
+				log.debug("Set focus failed.", exc_info=True)
+				return
+		review: textInfos.TextInfo = api.getReviewPosition()
+		# This script is available on the lock screen via getSafeScripts, as such
+		# ensure the review object does not contain secure information
+		# before displaying this object in braille line
+		if objectBelowLockScreenAndWindowsIsLocked(review.obj):
+			ui.reviewMessage(gui.blockAction.Context.WINDOWS_LOCKED.translatedMessage)
```
Comment (leonardder) [inline source/globalCommands.py:3576]: Avoid messages in this script

File: source/globalCommands.py [hunk 10]
```diff
@@ -3483,8 +3523,74 @@ def script_braille_scrollForward(self, gesture):
...
+				return
+		try:
+			review.updateCaret()
+		except NotImplementedError:
+			log.debug("Unable to move caret", exc_info=True)
+			# Translators: Reported when trying to move caret to the position
+			# of the review cursor but there is no caret.
+			ui.message(
+				_(
+					"No caret"
```
Comment (leonardder) [inline source/globalCommands.py:3591]: Avoid messages in this script

**Review by LeonarddeR — CHANGES_REQUESTED** [review]
While this looks interesting, I think the approach needs debate. 

1. I'd personally prefer to avoid introducing a new toggle as well as expanding the routing script like this. It is probably better to expand the routing behavior in the braille module instead.
2. Have you considered making the routing script support multiple key presses? You can route at first press, set focus ad second press, move system caret at third press or something similar.
3. Rather than copying most of the behavior of script_navigatorObject_moveFocus, please consider storing most of it logic in one or more helper methods and call these methods from both scripts to avoid duplicating code

**Comment by burmancomp** [comment]
Thank you @leonardder for your comments.

I selected new setting and gesture because I would like that user can decide how he/she likes that review and routing work.

I suppose that user wants to do the task with one press. If he/she wants to review with display without affecting system caret and route caret to current position if needed, he/she wants this with one press. This has been general behavior in screenreaders/braille display software for decades.

I agree with you that same code should be in one place. However maybe modifying of script_navigatorObject_moveFocus could be separate issue.

I'm unsure why messages should be avoided in this script although they are presented in braille as well.

Hopefully fixed "tethering automatically" problem. Also removed duplicate "braille.handler.routeTo(gesture.routingIndex)" line and changed doc strings to generic comments.

File: source/globalCommands.py (hunk 5 above)
Comment (burmancomp) [reply source/globalCommands.py:3529]: If pressing routing button causes message it appears again when dismissing with routing button. Reverting this at least for now.

**Comment by leonardder** [comment]
@burmancomp I saw some discussion took place in issue #14885 where @michaelDCurran provided three possible pathways to implement this, suggesting the second. It looks like you went for the first.
To be honest, I'd really want to take this back to the drawing board before it is wise going any further with this pr. It would be a pity if you'd invest more time in a solution that's eventually not going to pass

**Comment by burmancomp** [comment]
@leonardder I combined options 1 and 2. Option 2 would require extra step from user to get NVDA to follow review cursor. Putting options 1 and 2 together user can select if he/she wants to use automatic tether and take extra step or use tether to review, set "Route review cursor and system caret" to enabled and avoid that extra step. When reviewing text without affecting system caret and then routing caret to desired position has been in braille display softwares/screenreaders for decades. I think they have worked so that no extra step to route are needed.

What is the reason that there should be extra step? Why users who like to review text without affecting caret and then want to move caret should take that step?

Approach I'm using has no effect on user experience unless user wants it has. He/she can continue using braille as before but if he/she wants do things somewhat differently it would be possible. Selecting option 2 - if I understand it correctly - would mean that decision what users want/need would have been made for behalf of them but combining option 1 and 2 gives user possibility to decide.

**Comment by CyrilleB79** [comment]
I am not an expert in Braille nor a native English speaker, but I am wondering if there could not be a better phrasing for this option.

The option is called: "routeReviewCursorAndSystemCaret" and it can take the values "Enabled" and "Disabled".
* Is it clear that "Route" means "When pressing a routing key/cursor"?
* If the value is "Disabled" does it mean that neither the review cursor, nor the system caret is routed? Probably one of them is routed though.

Maybe a wording such as "Route system caret along with review cursor"?

Cc @XLTechie for your usual good English rewordings.

**Comment by leonardder** [comment]
> Maybe a wording such as "Route system caret along with review cursor"?

I came at `Move system caret when Routing review cursor`. I'm also struggling with the wording here.

**Comment by XLTechie** [comment]
@CyrilleB79 I am also not a braille user, and am not certain I understand this 
feature well enough. Here is a first attempt.

"When moving the review cursor with routing buttons, also move the system 
caret"?

Is that too long? Is it even true? :) Which came first? I am assuming review 
cursor, meaning that disable would only move that.

**Comment by burmancomp** [comment]
I changed it to "Route system caret along with review cursor" as @leonardder suggested.

**Comment by seanbudd** [comment]
Converting this to a draft until #15064 converges towards a solution, so we can decide on approach

**Comment by burmancomp** [comment]
> Converting this to a draft until #15064 converges towards a solution, so we can decide on approach

Appropriate and acceptable solution in this situation.

I am confident that @leonardder gets #15064 ready soon and then this can be closed. My only little concern is focusing objects. If user can focus objects when using speech output and keyboard, there should be same possibility with braille output and braille display buttons. Different users do/want to do things differently, and this should be taken into account. screenreader should give possibilities and leave decision to user.

**Comment by burmancomp** [comment]
This pull request is not needed anymore because #15064 is implementing all requested things.

**Pull request closed without merging**